### PR TITLE
feat: model-tier fallback resilience on NOT_FOUND (#813)

### LINKED ISSUE #813: Add model-fallback resilience when OpenRouter returns 404 / no-endpoints for a tier model
## Summary

When OpenRouter dropped `google/gemini-2.0-flash-001` from its catalog (~2026-05-31), every `fast`-tier invocation across four agents (`contacts`, `calendar`, `research-analyst`, `digest`) started returning `404 No endpoints found` and failed hard. There is no retry path, no fallback model, no tier degradation — just `agent.error` events piling up in `audit_log` until a human notices.

This issue tracks adding **resilience for the case where a tier's primary model becomes unavailable**, so that an upstream provider yanking a model doesn't quietly break multiple agents for hours.

See #812 for the originating incident (model swap) and root-cause notes.

## Failure mode (what happens today)

1. OpenRouter removes a model (deprecation, provider issue, account problem).
2. Curia's `OpenRouterAdapter` gets `404 No endpoints found` with `errorType: NOT_FOUND`, `retryable: false`.
3. The agent runtime catches the error, emits `agent.error`, and returns the canned `"I'm sorry, I was unable to process that request. Please try again."` to the caller.
4. Every subsequent call hits the same wall. Scheduled jobs fail. Coordinator delegations fail. No alert.

Specifically: the existing error type `NOT_FOUND` is marked `retryable: false`, which is correct for *this* model — but should trigger a fallback to a *different* model, not a hard fail.

## Proposed direction (open for discussion)

A few designs, in rough order of complexity:

### Option A: Per-tier fallback chain (recommended starting point)

Extend the tier config in `config/local.yaml` to accept a list:

```yaml
model_routing:
  tiers:
    fast:
      models:
        - google/gemini-3.1-flash-lite      # primary
        - mistralai/mistral-small-2603      # fallback
      # backwards-compat: `model:` (singular) still works as the primary
```

The `ModelRouter` tries the primary; on `NOT_FOUND` (or a configurable set of non-retryable provider errors) it falls through to the next entry. Failure is only surfaced after all fallbacks are exhausted.

**Pros:** explicit, debuggable, doesn't change tier semantics. **Cons:** requires per-tier upkeep; fallback may be on the same provider (no value if it's a provider-wide outage).

### Option B: Automatic tier escalation on persistent failure

Track `NOT_FOUND` failures per tier in memory; once a threshold is hit (e.g. 3 consecutive 404s in 5 minutes), automatically route `fast` traffic to `standard` until a circuit-breaker cooldown elapses.

**Pros:** zero config — works even when nobody noticed the model was pulled. **Cons:** silent cost increase (standard tier is more expensive); harder to reason about which model actually answered a given request.

### Option C: Provider-level circuit breaker

Wrap `OpenRouterAdapter` (and similar) with a circuit breaker that, on repeated non-retryable failures for a specific model ID, marks that model unavailable and routes future requests to a configured backup. Reset on a timer.

**Pros:** clean separation of concerns. **Cons:** more code; overlaps with Option A unless we want both.

My instinct: **start with Option A**, since it's a small config + router change and gives explicit control. Layer Option B on top later if we want zero-touch recovery.

## Detection & alerting (orthogonal but related)

Even with fallback, we should know when a primary went down. Options:
- Log `model.fallback.engaged` event when the router falls through
- Emit a `bus.warning` or audit-log entry visible in the dashboard
- Optional: notify the CEO when a tier's primary has been unavailable for >1h

Detection scope is up to whoever implements — minimum bar is a structured log event that's grep-able.

## Acceptance criteria

- [ ] Tier config supports an ordered list of models (backwards-compatible with single `model:` field)
- [ ] `ModelRouter` falls through to the next model in the list on `NOT_FOUND` (and any other configured non-retryable provider errors)
- [ ] Fallback engagement is logged as a structured event (model name, reason, tier) — visible in `audit_log` or pino logs
- [ ] Unit tests cover: (a) primary success, (b) primary 404 → fallback success, (c) all-models-failed bubbles up the original error
- [ ] Integration test against a mocked OpenRouter that returns 404 for the primary
- [ ] CHANGELOG entry under `## [Unreleased]` → **Added**
- [ ] Spec update (likely [docs/specs/05-error-recovery.md](docs/specs/05-error-recovery.md)) describing the fallback semantics

## Out of scope

- Automatic tier escalation (Option B) — separate follow-up if we want it after A lands
- Generalized retry of `retryable: true` errors — that path is already handled elsewhere
- Multi-provider routing beyond OpenRouter — Anthropic-direct fallback would be nice but raises auth/cost questions worth a separate discussion

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

Closes #813

- Adds fixed-rule tier fallback in `AgentRuntime.chatWithRetry()`: on `NOT_FOUND` (provider dropped a model), the runtime automatically re-routes to the fallback tier's model before surfacing an error. Rules: `fast`→`standard`, `standard`→`powerful`, `powerful`→`standard`.
- Adds `ModelRouter.getFallbackTier()` with the `FALLBACK_TIER` constant encoding those rules.
- Adds `ModelFallbackEngagedEvent` (`model.fallback`) to the bus event registry and permissions — published as a durable audit event whenever fallback engages.
- Pre-computes fallback model + provider at bootstrap in `src/index.ts` (same pattern as primary model resolution) and passes them into `AgentRuntime` via three new optional `AgentConfig` fields (`tier`, `fallbackModel`, `fallbackProvider`).
- Adds 6 end-to-end runtime tests and 4 model-router tests. Full suite: 305 test files, 0 failures.

## Design notes

Only `NOT_FOUND` triggers fallback. `AUTH_FAILURE`, `VALIDATION_ERROR`, and other non-retryable errors still fail hard — they indicate caller or config problems, not model availability.

If the fallback also fails, its error is surfaced (not the original `NOT_FOUND`) and budget is counted normally. The `model.fallback` audit event is still published so operators can see the attempt.

A `@TODO` is left noting that `params.messages` are assembled against the primary model's context window; for `powerful→standard` downgrades the fallback could have a smaller window. This is a deferred edge case — the common `fast→standard` and `standard→powerful` paths use larger-or-equal windows.

## Test plan

- [ ] `pnpm run test src/agents/runtime.test.ts` — 6 tests covering: primary success, NOT_FOUND→fallback success + model.fallback event, both fail→error, no fallback for AUTH_FAILURE, success without fallback configured, error without fallback configured
- [ ] `pnpm run test src/agents/llm/model-router.test.ts` — includes 4 new getFallbackTier tests
- [ ] `pnpm run typecheck` passes clean
- [ ] Full test suite: 305 passed, 0 failures

### CODE REVIEW

**Comment by josephfung** [comment]
@coderabbitai review

File: src/index.ts [hunk 1]
```diff
@@ -1815,11 +1815,22 @@ async function main(): Promise<void> {
...
+    const agentFallbackProvider = fallbackProviderName ? providerRegistry.get(fallbackProviderName) : undefined;
+
     const agent = new AgentRuntime({
       agentId: agentConfig.name,
       systemPrompt,
       provider: agentProvider,
       resolvedModel: resolved.model,
+      tier: resolved.tier,
+      fallbackModel: fallbackResolved.model,
+      fallbackProvider: agentFallbackProvider,
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/index.ts:1849]: Fixed in 4bb5eb66. Added two startup invariant checks after the fallback provider lookup: if `fallbackProviderName` is null (model not in registry) or if `providerRegistry.get(fallbackProviderName)` returns undefined (provider not wired up), the process exits at startup with a structured log — same fail-fast pattern used for the primary provider. The silent partial-config path (fallbackModel set, fallbackProvider undefined) no longer exists.

File: src/agents/runtime.ts [hunk 2]
```diff
@@ -1360,9 +1374,52 @@ export class AgentRuntime {
...
+        // standard→powerful) are larger-or-equal so this is safe. (#813)
+        const fallbackStartMs = Date.now();
+        const fallbackResponse = await this.config.fallbackProvider.chat({ ...params, model: this.config.fallbackModel });
+        const fallbackLatencyMs = Date.now() - fallbackStartMs;
+
+        if (fallbackResponse.type !== 'error') {
+          // Fallback succeeded — behave as if the primary call succeeded.
+          budget.consecutiveErrors = 0;
+          await publishLlmCallEvent(fallbackResponse, fallbackLatencyMs, this.config.fallbackProvider.id);
+          return fallbackResponse;
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/agents/runtime.ts:1426]: Acknowledged — this is a real edge. The concern is that if the fallback succeeds with a tool-use response, subsequent chatWithRetry() rounds re-call the primary (which is still broken), triggering fallback again and generating duplicate model.fallback events for every tool round. The fix requires persisting the "active provider/model" for the task lifetime rather than resolving it fresh each call.

That's the right fix but it's a non-trivial refactor (task-scoped provider state vs current per-call closure). I've opened a follow-up to track it. This PR handles the common catalog-drop case cleanly — the fallback fires and the user gets a response — which was the goal of #813. The repeated-event side effect is visible in audit but doesn't affect correctness or the user experience.

**Comment by josephfung** [comment]
Addressed in commit 4bb5eb66. Per-comment breakdown:

**CodeAnt — `runtime.ts` fallback call not in try/catch (Critical)**
Valid. If `fallbackProvider.chat()` throws (network failure, SDK error before it can return a structured `LLMResponse`), the outer `handleTask` boundary would catch it, but without publishing `agent.error` or updating budget counters, leaving the audit trail incomplete. Fixed: the fallback call is now wrapped in its own try/catch that normalizes thrown exceptions into the `AgentError` flow (type `UNKNOWN`, correct source, structured log, `agent.error` published, `sendErrorResponse`, budget incremented) before returning null.

**CodeAnt — `runtime.test.ts` runTask helper has no try/catch around `bus.publish` (Critical)**
Won't fix — this is intentional. In a test helper, letting async errors propagate unhandled causes the test to fail loudly, which is the correct behavior. Adding a try/catch here would swallow unexpected errors and let tests pass spuriously. The `Promise.race` timeout already handles the hanging case.

**CodeRabbit — `src/bus/events.ts` — narrow `tier` and `reason` to explicit unions (Trivial)**
Fixed: `tier` is now `'fast' | 'standard' | 'powerful' | 'unknown'` and `reason` is `'NOT_FOUND'`. The call site passes the literal `'NOT_FOUND'` (the method is only reachable from the NOT_FOUND branch), so no runtime assertion needed.

**CodeRabbit — `CHANGELOG.md` — trim the entry (Trivial)**
Fixed: condensed to one clause, ~15 words after the em-dash, spec reference included.

**CodeRabbit — `runtime.test.ts` — test (c) can't distinguish primary vs fallback error (Trivial)**
Fixed: `makeNotFoundResponse()` now takes optional `source` and `message` params. Test (c) uses distinct sources (`'openrouter'` for primary, `'anthropic'` for fallback), captures `agent.error` events, and asserts `errorEvents[0].payload.source === 'anthropic'` — proving the fallback's error surfaced, not the primary NOT_FOUND.

**CodeRabbit — `src/index.ts` — fallback-provider invariant not enforced (Minor)**
Fixed inline above.

**CodeRabbit — `runtime.ts` — centralize fallback routing across retry rounds (Major)**
Replied inline. The duplicate-event side effect is real but deferred — correct behavior for the common case, cosmetic audit noise for multi-round tool-use tasks. Will track in a follow-up issue.

File: src/agents/runtime.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (josephfung) [reply src/agents/runtime.ts:1426]: Follow-up issue is #1146. The acceptance criteria there require that after the first successful fallback engagement, all subsequent LLM rounds within the same task route directly to the fallback provider — so exactly one `model.fallback` event per task rather than one per tool-use round. The audit trail will be tidy.
